### PR TITLE
lxml recipe

### PR DESCRIPTION
**Pull request by goffi-contrib**
This PR fixes the small issues found in https://github.com/kivy/python-for-android/pull/1166.
It's a working recipe for lxml and its dependencies.

### CODE REVIEW

File: pythonforandroid/recipes/levenshtein/__init__.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    version = '0.12.0'
+    url = 'https://pypi.python.org/packages/source/p/python-Levenshtein/python-Levenshtein-{version}.tar.gz'
+    depends = [('python2', 'python3'), 'setuptools']
+
+    call_hostpython_via_targetpython = False
+
+    def get_recipe_env(self, arch):
+        env = super(LevenshteinRecipe, self).get_recipe_env(arch)
+        libxslt_recipe = Recipe.get_recipe('libxslt', self.ctx)
+        libxml2_recipe = Recipe.get_recipe('libxml2', self.ctx)
```
Comment (AndreMiras) [inline pythonforandroid/recipes/levenshtein/__init__.py]: I'm not sure how can these be working as `Recipe` is undefined :thinking:

**Comment by goffi-contrib** [comment]
Hello @AndreMiras thanks for feedback. This PR is just a small correction over https://github.com/kivy/python-for-android/pull/1166, and I've kept the original commits. The original author is @Scefing, I'll wait to see if she is still interested in this patch, else I'll clean it up myself.

**Comment by goffi-contrib** [comment]
@ZachGoldberg @Scefing just pinging you to see if you wanna continue on this recipe :)

**Comment by goffi-contrib** [comment]
@AndreMiras I've updated the PR, as @ZachGoldberg and @Scefing don't look active anymore.
I've squashed initial commits in a first one, then a second one with my fixes. I've kept separate commits because of different authors.

About the undefined Recipe you have found in levenshtein recipe, I've fixed it. But I'm not sure what this recipe is used for, as it's not linked in lxml recipe or any dependency.

Anyway, this is hopefully OK now, I've tested it working OK 2 days ago. Thanks.

**Comment by goffi-contrib** [comment]
@AndreMiras yes I agree. I'll remove it tonight (CET), I've started my working day now :)

**Comment by AndreMiras** [comment]
Great, can't wait for it :+1: 
And the `levenshtein` recipe is indeed failing, but straight from the import:
```
Traceback (most recent call last):
  File "setup_testapp_python3.py", line 30, in <module>
    package_data={'testapp': ['*.py', '*.png']}
  File "/usr/lib/python2.7/distutils/core.py", line 151, in setup
    dist.run_commands()
  File "/usr/lib/python2.7/distutils/dist.py", line 953, in run_commands
    self.run_command(cmd)
  File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
    cmd_obj.run()
  File "/app/pythonforandroid/bdistapk.py", line 81, in run
    main()
  File "/app/pythonforandroid/toolchain.py", line 999, in main
    ToolchainCL()
  File "/app/pythonforandroid/toolchain.py", line 532, in __init__
    getattr(self, args.subparser_name.replace('-', '_'))(args)
  File "/app/pythonforandroid/toolchain.py", line 145, in wrapper_func
    build_dist_from_args(ctx, dist, args)
  File "/app/pythonforandroid/toolchain.py", line 166, in build_dist_from_args
    = get_recipe_order_and_bootstrap(ctx, dist.recipes, bs)
  File "/app/pythonforandroid/graph.py", line 112, in get_recipe_order_and_bootstrap
    name, ctx, orders=new_possible_orders)
  File "/app/pythonforandroid/graph.py", line 35, in recursively_collect_orders
    recipe = Recipe.get_recipe(name, ctx)
  File "/app/pythonforandroid/recipe.py", line 614, in get_recipe
    mod = import_recipe('pythonforandroid.recipes.{}'.format(name), recipe_file)
  File "/app/pythonforandroid/recipes/levenshtein/__init__.py", line 1, in <module>
    from pythonforandroid.toolchain import CompiledComponentsPythonRecipe
ImportError: cannot import name CompiledComponentsPythonRecipe
```
Full log: https://api.travis-ci.org/v3/job/446032065/log.txt

**Comment by goffi-contrib** [comment]
@AndreMiras I've removed levenshtein recipe, the CI is failing because of name resolution issue with gradle (maybe the server was down during the test?). We would need to restart it, but I'm not sure how I can do (I can still do a minor change to PR, but it would be better to just relaunch travis).

**Comment by goffi-contrib** [comment]
I've closed/reopened PR to trigger the travis build

File: pythonforandroid/recipes/lxml/__init__.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
+from pythonforandroid.toolchain import Recipe, shutil
+from pythonforandroid.recipe import CompiledComponentsPythonRecipe
+from os.path import exists, join, dirname
+
+
+class LXMLRecipe(CompiledComponentsPythonRecipe):
+    version = "3.6.0"
+    url = "https://pypi.python.org/packages/source/l/lxml/lxml-{version}.tar.gz"
+    depends = ["python2", "libxml2", "libxslt"]
```
Comment (AndreMiras) [inline pythonforandroid/recipes/lxml/__init__.py]: So you have only tested it under `python2`?

Comment (goffi-contrib) [reply pythonforandroid/recipes/lxml/__init__.py]: yes, my project is not yet Python 3 compatible.

**Comment by AndreMiras** [comment]
Yes Travis is flaky often because of the network. But I usually restart the failing jobs myself to check their outcome.
The build I was most interested in is failing for different reasons however: https://api.travis-ci.org/v3/job/446531236/log.txt 
Most interesting part of the log is here for records https://pastebin.com/KKE0b6jq
Basically the run command was:
```
  RAN: /app/venv/bin/python setup_testapp_python3.py apk --sdk-dir /opt/android/android-sdk --ndk-dir /opt/android/crystax-ndk --bootstrap sdl2 --requirements libxml2,lxml,python3crystax,libxslt
```
Also see:
```
('recipes modified:', set([u'libxml2', u'lxml', u'libxslt']))
('recipes to build:', set([u'libxml2', u'lxml', u'libxslt']))
('requirements:', u'libxml2,lxml,python3crystax,libxslt')
```
However your recipe seems to be `python2` only which conditional builds doesn't support currently, see https://github.com/kivy/python-for-android/blob/b6ca832/ci/rebuild_updated_recipes.py
Do you know if your recipe could be working under python3crystax?

File: pythonforandroid/recipes/lxml/__init__.py (hunk 2 above)
Comment (AndreMiras) [reply pythonforandroid/recipes/lxml/__init__.py]: OK so it means, I either need to test it by hand, or try to make conditional builds also working with `python2` :smile:

**Comment by goffi-contrib** [comment]
@AndreMiras I don't know about python3crystax, as I've said earlier, I've just fixed recipe from an other PR which is inactive now. I don't see why it would not work, but it would take too much time for me to try on python 3 now, I haven't python 3 app to test and I'm working in parallel on other tasks.

**Comment by AndreMiras** [comment]
No problem :+1: 
Just that I need to take a deeper look since some part of the CI would not be of any help.
So that will just require a bit more time

**Comment by goffi-contrib** [comment]
@AndreMiras most test are passing, beside the one with python3crystax. The other failing one is due to docker and smells like an other network issue (it can't find p4a image).
I'll try to add python3crystax in `depends` to see where it goes

**Comment by goffi-contrib** [comment]
> Yes other tests are passing because they try to compile core recipes and you didn't touch core recipes.
> That's why I'm mainly interested in the conditional build because it's the one that tries to build only the recipes that got touched.
> But since it's failing I'll need to test your pull request manually to make sure things are OK

OK, I don't the PR for now then.

**Comment by AndreMiras** [comment]
You though I already forgot you? Nope :smile:
I hate exhausting my laptop trying to compile ton of recipes so I made Travis deal with it for us :wink: 
In https://github.com/kivy/python-for-android/pull/1435 we're now handling `python2` target (as a fallback).
Once/if it gets merged I can merge your PR. Well I will kinda give it a rebase/retry on my fork first.

**Comment by goffi-contrib** [comment]
> You though I already forgot you? Nope smile
> I hate exhausting my laptop trying to compile ton of recipes so I made Travis deal with it for us wink
> In #1435 we're now handling `python2` target (as a fallback).
> Once/if it gets merged I can merge your PR. Well I will kinda give it a rebase/retry on my fork first.

he he, thanks for the update :)

**Comment by AndreMiras** [comment]
Great, seems like Travis is happy :+1: 
Thanks!

**Pull request merged**